File: cmd/fail.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/Pernek-Enterprises/dispatch/internal/config"
	"github.com/Pernek-Enterprises/dispatch/internal/pipe"
)

func Fail(args []string) {
	message := joinArgs(args)
	if message == "" {
		fmt.Fprintln(os.Stderr, "dispatch: usage: dispatch fail \"reason\"")
		os.Exit(1)
	}

	jobID := os.Getenv("DISPATCH_JOB_ID")
	taskID := os.Getenv("DISPATCH_TASK_ID")

	// Write failure result
	if jobID != "" {
		resultPath := filepath.Join(config.Root, "jobs", "active", jobID+".result.md")
		os.WriteFile(resultPath, []byte("FAILED: "+message+"\n"), 0644)
	}

	pipePath := getPipePath()
	err := pipe.Send(pipePath, pipe.Message{
		Type:   "fail",
		JobID:  jobID,
		TaskID: taskID,
		Reason: message,
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "dispatch: failed to notify foreman: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("✗ Step failed: %s\n", message)
}

File: cli/fail.ts
import { loadConfig } from "../config.js";
import { sendPipe } from "../pipe.js";

export function failCmd(args: string[]): void {
  let jobId = process.env.DISPATCH_JOB_ID ?? "";
  const rest: string[] = [];

  for (let i = 0; i < args.length; i++) {
    if ((args[i] === "--job" || args[i] === "-j") && args[i + 1]) { jobId = args[++i]; }
    else { rest.push(args[i]); }
  }

  if (!jobId) { console.error("dispatch: --job is required"); process.exit(1); }
  const reason = rest.join(" ");
  if (!reason) { console.error("dispatch: reason is required"); process.exit(1); }

  const cfg = loadConfig();
  sendPipe(cfg.pipePath, { type: "fail", jobId, reason });
  console.log(`✓ Failure reported for job ${jobId}`);
}
